File: practitioner-code/payments-java/src/main/java/com/example/payments/domain/WalletId.java
package com.example.payments.domain;

import java.util.Objects;
import java.util.UUID;

/**
 * Typed identifier (a "tiny type"). NEVER pass raw UUID/String/long for ids
 * across the domain - it lets you swap a customerId for a walletId silently.
 * A WalletId can ONLY be a WalletId; the compiler stops the bug.
 *
 * Cost: a class. Benefit: a whole bug class becomes impossible.
 */
public record WalletId(UUID value) {
    public WalletId {
        Objects.requireNonNull(value, "value");
    }

    public static WalletId newId() {
        return new WalletId(UUID.randomUUID());
    }
}

File: practitioner-code/payments-python/src/payments/domain/wallet_id.py
"""
Typed identifier (a "tiny type"). NEVER pass raw UUID/str across the domain -
it lets you swap a customer_id for a wallet_id silently. A WalletId can ONLY
be a WalletId; mypy and runtime equality both stop the bug.

(typing.NewType would help static checkers but not runtime - a frozen
dataclass is a real distinct type that also enables structured comparison.)
"""
from __future__ import annotations

from dataclasses import dataclass
from uuid import UUID, uuid4


@dataclass(frozen=True, slots=True)
class WalletId:
    value: UUID

    @classmethod
    def new(cls) -> "WalletId":
        return cls(uuid4())

    def __str__(self) -> str:
        return str(self.value)
